hal/device.c

```diff
@@ -0,0 +1,49 @@
+#include "device.h"
+
+static struct device_type *devices;
+
+static struct device_type **find_device(const char *name, unsigned len) {
+  struct device_type **p;
+  for (p = &devices; *p; p = &(*p)->next)
+    if (strlen((*p)->name) == len && strncmp((*p)->name, name, len) == 0) break;
```

Comment:
Change from
`strncmp((*p)->name, name, len) == 0`
to
`!strncmp((*p)->name, name, len)`